src/client_builder.rs

```diff
@@ -1,55 +1,159 @@
-use crate::{Client, Result};
+use crate::{Client, Connection, Error, Result};
+
+use lazy_static::lazy_static;
 use std::io::{Read, Write};
 use std::net::TcpStream;
 
 #[cfg(feature = "native-tls")]
-use native_tls::{TlsConnector, TlsStream};
+use native_tls::TlsConnector as NativeTlsConnector;
+
+use crate::extensions::idle::SetReadTimeout;
+#[cfg(feature = "rustls-tls")]
+use rustls_connector::{
+    rustls,
+    rustls::{Certificate, ClientConfig, RootCertStore, ServerName},
+    rustls_native_certs::load_native_certs,
+    RustlsConnector,
+};
+#[cfg(feature = "rustls-tls")]
+use std::sync::Arc;
+
+#[cfg(feature = "rustls-tls")]
+struct NoCertVerification;
+
+#[cfg(feature = "rustls-tls")]
+impl rustls::client::ServerCertVerifier for NoCertVerification {
+    fn verify_server_cert(
+        &self,
+        _: &Certificate,
+        _: &[Certificate],
+        _: &ServerName,
+        _: &mut dyn Iterator<Item = &[u8]>,
+        _: &[u8],
+        _: std::time::SystemTime,
+    ) -> std::result::Result<rustls::client::ServerCertVerified, rustls::Error> {
+        Ok(rustls::client::ServerCertVerified::assertion())
+    }
+}
+
 #[cfg(feature = "rustls-tls")]
-use rustls_connector::{RustlsConnector, TlsStream as RustlsStream};
+lazy_static! {
+    static ref CACERTS: RootCertStore = {
+        let mut store = RootCertStore::empty();
+        for cert in load_native_certs().unwrap_or_else(|_| vec![]) {
+            if let Ok(_) = store.add(&Certificate(cert.0)) {}
+        }
+        store
+    };
+}
+
+lazy_static! {
+    static ref STARTLS_CHECK_REGEX: regex::bytes::Regex =
+        regex::bytes::Regex::new(r"\bSTARTTLS\b").unwrap();
+}
+
+/// The connection mode we are going to use
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum ConnectionMode {
+    /// Automatically detect what connection mode should be used.
+    ///
+    /// This will use TLS if the port is 993, and otherwise STARTTLS if available.
+    /// If no TLS communication mechanism is available, the connection will fail.
+    #[cfg(any(feature = "native-tls", feature = "rustls-tls"))]
```

Comment:
I think this needs to not be `cfg`-guarded since we now always set it in `ClientBuilder::new` (in fact the code probably won't compile as-is).

Reply:
Should we just make the otherones also not Cfg guarded since we have the proper error code response now when no Tls is compiled in?

Reply:
No, let's keep those guarded (we can always un-guard them now that it's `#[non_exhaustive]`). It'll still then serve as a signal to folks to remember to enable the TLS features if they specifically try to enforce TLS.